internal/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,260 @@
+package ratelimit
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/go-resty/resty/v2"
+)
+
+type RateLimitInfo struct {
+	Limit     int // x-ratelimit-limit
+	Period    int // x-ratelimit-period
+	Remaining int // x-ratelimit-remaining
+	Reset     int // x-ratelimit-reset (seconds until reset)
+}
+
+func (r *RateLimitInfo) IsNearLimit(threshold float64) bool {
+	if r.Limit == 0 {
+		return false
+	}
+	return float64(r.Remaining)/float64(r.Limit) < threshold
+}
+
+func (r *RateLimitInfo) ShouldThrottle(threshold float64) bool {
+	return r.IsNearLimit(threshold)
+}
+
+type Manager struct {
+	mu                 sync.RWMutex
+	rateLimitInfo      *RateLimitInfo
+	enabled            bool
+	threshold          float64
+	activeRequests     int
+	requestSemaphore   chan struct{}
+	lastRequestTime    time.Time
+	minRequestInterval time.Duration
+	debug              bool
+}
+
+func NewManager() *Manager {
+	return &Manager{
+		enabled:            os.Getenv("PORT_RATE_LIMIT_DISABLED") == "",
+		threshold:          0.02,
+		debug:              os.Getenv("PORT_DEBUG_RATE_LIMIT") != "",
+		requestSemaphore:   make(chan struct{}, 50),
+		minRequestInterval: 50 * time.Millisecond,
+	}
+}
+
+func (m *Manager) GetInfo() *RateLimitInfo {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	if m.rateLimitInfo == nil {
+		return nil
+	}
+
+	return &RateLimitInfo{
+		Limit:     m.rateLimitInfo.Limit,
+		Period:    m.rateLimitInfo.Period,
+		Remaining: m.rateLimitInfo.Remaining,
+		Reset:     m.rateLimitInfo.Reset,
+	}
+}
+
+func (m *Manager) SetEnabled(enabled bool) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.enabled = enabled
+}
+
+func (m *Manager) SetThreshold(threshold float64) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if threshold >= 0.0 && threshold <= 1.0 {
+		m.threshold = threshold
+	}
+}
+
+func (m *Manager) log(logString string) {
+	if m.debug {
+		fmt.Print(logString)
+	}
+}
```

Comment:
Instead of this, pass a [`slog.Logger`](https://pkg.go.dev/log/slog#Logger) or a [`log.Logger`](https://pkg.go.dev/log#Logger) and bake the `debug` boolean by setting the level (or setting the output file to dev/null in the `log.Logger`).

Reply:
This will also simplify the calls to logs from stuff like:

```go
m.log(fmt.Sprintf("DEBUG: RequestMiddleware called, enabled: %v\n", m.enabled))
```

to:

```go
m.logger.Debug("RequestMiddleware called", "enabled", m.enabled)
```

Reply:
You can even set attributes like `enabled` on the logger, so all logs will print it.